### ISSUE TITLE
Package the exe as a self contained app so it can run cross platform with being re-built

### ISSUE THREAD

**Issue by rfennell** [MaxMelcher/AzureDevOps.WikiPDFExport#55]
As you might remember I have an [Azure DevOps extension that wrappers this tool](https://marketplace.visualstudio.com/items?itemName=richardfennellBM.BM-VSTS-WikiPDFExport-Tasks).

This task runs fine on Windows based build agents, but I cannot get it to work on Linux and MacOS based ones using the EXE you provide in the release. 

![image](https://user-images.githubusercontent.com/324718/111628470-281bb900-87e8-11eb-838d-8b92c62eead2.png)

Given the error message, I think this is because the EXE is not built as a self-contained application. 

Is this something
1. You have tested
2. Would consider packaging the EXE as in the release so the task can pull down the current version in a form that can be run on all platforms?

**Comment by MaxMelcher** [comment]
Hey Richard - sure! 

I want to build & publish the converter with GitHub actions soon - doing a second build for a linux runtime should be easy. 
Not sure when I can do it, but yes thats something I would love to see. 

1. tested - no; the path and downloading could be different on linux. 
2. consider - yes

PR welcome :)

**Comment by rfennell** [comment]
I have done some initial research.

As you already have a single file windows distribution, I have focused on creating a Linux one.

To do this I used the command ( not 100% sure over the flag, see [this post on errors seen if p:PublishReadyToRun=true](https://github.com/dotnet/docs/issues/17956)

`dotnet publish -r linux-x64 -p:PublishSingleFile=true -p:PublishReadyToRun=false`

This resulted in a 122Mb file. I copied this to the Ubuntu VM and tried to run it, but it failed with a

> bash: ./azuredevops-export-wiki; Permission denied

This was fixed with the command

`chmod u+x azuredevops-export-wiki`

The executable now ran, it showed the QT warning so it did start. But them gave an error

![image](https://user-images.githubusercontent.com/324718/111657919-c9fccf00-8803-11eb-99a9-f96887164dd1.png)

I tried copying over the libwkhtmltox.dll to the same folder. it made no difference.

I think the issue is related to the `libwkhtmltox` assembly being added to the project as a prebuilt assembly. Either it is not being packaged up correctly, or is built in the wrong format.

I will continue to work on this as time allows

**Comment by rfennell** [comment]
A bit more of an update, tried exactly the same publish command with but with a simple 'Hello World' console application. I was able to build a single file for Windows and another for Linux. These both ran without error on their appropriate platform.

So this does point to the issue being related `libwkhtmltox` assembly as I mentioned above. 

Given this, I am not sure I can progress this much more while this assembly is loaded in this manner.

**Comment by MaxMelcher** [comment]
We could load it dynamically - download if not present. This would bring down the exe size a lot.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by rfennell** [comment]
I think the issue could be that the [wkhtmlopdf library](https://wkhtmltopdf.org/downloads.html) is platform specific. So you can't load the Windows version on Linux, and I assume MacOS. No sure how to resolve this as I have no experience of writing dotnet core apps for Linux

**Comment by MaxMelcher** [comment]
@rfennell  the latest version runs on ubuntu - how to best provide this to your Azure DevOps task? 
89f5399583a8f22de5ee55ddb50ba38fa4270800

**Comment by rfennell** [comment]
It may 'just' work, I will test it and report back

**Comment by MaxMelcher** [comment]
I briefly scrolled through your code to see how it works - it pulls the exe file from the release. That's fine for windows, I guess. 

Should I just add a Linux binary without .exe extension to the release to make it easier for you?

**Comment by rfennell** [comment]
After a bit of testing, the issue I see is that on non-windows (I was using Ubuntu) you can't package or run the tool as an EXE.If you try to run the exe you get

![image](https://user-images.githubusercontent.com/324718/132223203-be738133-be4f-4ff2-8412-d110fa453c4c.png)

So I  need to run the command 

`dotnet AzureDevOpsPDFExport.dll`. 

If I run just the above command line, the tool runs and exits as expected. However, to be useful I need to pass in command line parameters and I can't get this to work. I have tried to use the `--` (as [detailed here](https://github.com/dotnet/sdk/issues/5097)) but it has no effect, I get the out of range exception.

![image](https://user-images.githubusercontent.com/324718/132223367-d0942b99-a090-4034-bc59-f183b1c8d1da.png)

## Packaging
But that can all be worked out. The major point is that within the release we also need the DLL, as well as the Windows EXE. I will then alter my task to run the EXE for windows  or the `dotnet AzureDevOpsPDFExport.dll` for other platforms.

The only alternative is for me to build the DLL on the fly, but this requires the DOTNET SDK is on the build agent, which though likely is not guaranteed.

## Timescale
A final question, what is the timescale to release this new version. At present my task does not support pre-view releases, I have modified my private version so it can pull down the new version as a test. If the cross-platform version is going to be a while, I will add the option to use pre-release versions via a new task parameter. If it is about to come out, I will no bother, I will just wait

**Comment by rfennell** [comment]
Sorry, only just re-read your last comment, yes, adding the DLL as well as the EXE to the release package, would really help.

Also on the parameter front, the problem appears to be `-p` folder option. If I miss this out it work

![image](https://user-images.githubusercontent.com/324718/132311437-7271c5b7-c783-4bce-990a-8698f00aa1f3.png)

If you add the `-p` to a valid folder it does not error, but is ignored

![image](https://user-images.githubusercontent.com/324718/132311219-efb474d3-d9a5-4d9f-96ca-c048a53d23d8.png)

I will try building the DLL locally and see if I can spot the issue

**Comment by rfennell** [comment]
@MaxMelcher  I now have a private version of my task that builds the correct command line for Linux/Mac and also  provides a flag to allow the us of pre-releases of the tool. Should work as soon as the DLL is present in the  release package

**Comment by MaxMelcher** [comment]
.NET 6 release is in 9 days - I think it makes sense to release afterwards so we do not have two breaking changes. 
@rfennell do you support .NET 6?

**Comment by rfennell** [comment]
Makes perfect sense to wait. 

My task is written in Node.js so as long as the agent has a suitable ru time installed all should be good

**Comment by MaxMelcher** [comment]
Thanks for confirming! I will release a beta next week to see if NET6 breaks anything.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by MaxMelcher** [comment]
Just release the v4 version @rfennell - can you please test?

**Comment by rfennell** [comment]
I am on holiday this week
, but will check it as soon as I get back at the end of the week

Just to check the new release contains the exe, is this now OK for linux/mac, you used to need a dell for these?

**Comment by MaxMelcher** [comment]
the executable for linux does not have an extension: https://github.com/MaxMelcher/AzureDevOps.WikiPDFExport/releases/download/4.0.0/azuredevops-export-wiki

**Comment by rfennell** [comment]
Just released V2 of my [extension](https://marketplace.visualstudio.com/items?itemName=richardfennellBM.BM-VSTS-WikiPDFExport-Tasks). It is tested on Windows and Linux. I can't see any release it should not work on Mac too if we had a suitably packaged executable.

The only changes I had to make were
- Alter the task to generate the correct command line to run for Windows and Linux
- Run CHMOD on Linux to set the execution permission

I need to make no changes for your updated command line parameter (header etc) as I just pass this information through as string parameter.

The only gotcha is the need for .NET6 on the agent. In many cases this runtime will be present on a build agent e.g if VS2022 is installed. If it is not present it can be installed as a step in a build pipeline using the [usedotnet task](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/tool/dotnet-core-tool-installer?view=azure-devops) before this task is called.

> Note: this technique will work for both V1 and V2 of this task

```
steps:
- task: UseDotNet@2
  displayName: 'Use .NET Core runtime'
  inputs:
    packageType: 'runtime'
    version: '6.0.x'
    includePreviewVersions: false
- task: richardfennellBM.BM-VSTS-WikiPDFExport-Tasks.WikiPDFExportTask.WikiPdfExportTask@2
  displayName: 'Export Single File'
  inputs:
    cloneRepo: false
    usePreRelease: false
    localpath: '$(System.DefaultWorkingDirectory)'
    singleFile: 'infile.md'
    outputFile: '$(Build.ArtifactStagingDirectory)/singleFile.pdf'
    ...
``` 

Thanks @MaxMelcher for all your work on getting this tool ports to .NET6

**Comment by MaxMelcher** [comment]
> I can't see any release it should not work on Mac too if we had a suitably packaged executable.

I cant test the Mac version - can you? How should the extension be?
I release with GitHub Actions now, so adding this should be easy :)

**Comment by rfennell** [comment]
I have a couple if options. 

There are a couple of Macs the office used for testing. Just need to pick one up next time I go in

The other option is I just but test on a hosted Azure DevOps build agent.

I will get back to you

**Comment by rfennell** [comment]
@MaxMelcher,  good news, as suspected it works fine on a Mac. I tested on an old Mac Mini running macOS High Sierra 10.13.6.

I tested both using `dotnet run` (after cloning the repo) and packaged up after running `dotnet build -- runtime osx-x64 --self-contained`; both worked as expected.

Not sure how you want to add the Mac packaged version to release? 

Once, I know I can alter my task to download the correct version. At present, if not on Windows, it tries to use the Linux build, which won't work on the Mac (I tried you get a `cannot execute binary file` error which is not unreasonable)

**Comment by MaxMelcher** [comment]
ok thats easy. I just publish the third binary and add -osx to the filename. 
https://github.com/MaxMelcher/AzureDevOps.WikiPDFExport/releases/tag/4.0.3

Ok?

**Comment by rfennell** [comment]
Perfect

**Comment by rfennell** [comment]
Just released 2.1.2 of my extension that now fully support Macs as well as Windows and Linux

**Comment by MaxMelcher** [comment]
Very cool - thanks! :)
